### PR TITLE
Fix iteration number tracking with pipeline.reset

### PR DESCRIPTION
**Pull request by mzient**
#### Why we need this PR?
*Pick one, remove the rest*
- It fixes bugs:
   - reset schedules additional processing
   - data race when Reset is called
   - batch index is ambiguous for small batches

#### What happened in this PR?
*Fill relevant points, put NA otherwise. Replace anything inside []*
 - What solution was applied:
     * Batches are scheduled to be placed in a specific queue entry
     * Entries are bound to a particular global iteration number (not withing epoch and not a job subscription number)
     * Jobs are scheduled with a unique ID instead of batch index
 - Affected modules and functionalities:
     * External source
 - Key points relevant for the review:
     * N/A
 - Validation and testing:
     * Existing tests apply, added a test with large data
 - Documentation (including examples):
     * N/A

**JIRA TASK**: DALI-1910

Co-authored-by: Krzysztof Lecki <klecki@nvidia.com>

### CODE REVIEW

File: dali/python/nvidia/dali/external_source.py [hunk 1]
```diff
@@ -97,32 +100,47 @@ def __init__(
...
         return (arg,)
 
     def reset_indices(self):
         self.current_iter = 0
         self.current_sample = 0
+        self.cancel_prefetch()
 
-    def schedule_batch(self, pool, context_i, batch_size):
+    def cancel_prefetch(self):
+        self.scheduled_ahead = 0
```
Comment (JanuszL) [inline dali/python/nvidia/dali/external_source.py:128]: Does that cancel prefetch (I don't see how `self.scheduled_ahead = 0` does that)?

**Review by JanuszL — APPROVED** [review]

File: dali/test/python/test_external_source_parallel_utils.py [hunk 2]
```diff
@@ -142,6 +142,7 @@ def check_stop_iteration_resume(pipe, batch_size, layout):
         try:
             while True:
                 (r, ) = pipe.run()
+                r = [r.at(i) for i in range(len(r))]
                 output.append(r)
```
Comment (klecki) [inline dali/test/python/test_external_source_parallel_utils.py:146]: I think we should do a copy here, for example:
```suggestion
                output.append(np.copy(r.as_array()))
```

**Comment by mzient** [comment]
!build

**Review by klecki — APPROVED** [review]

File: dali/python/nvidia/dali/external_source.py (hunk 1 above)
Comment (mzient) [reply dali/python/nvidia/dali/external_source.py:128]: This is sort of spread between pool.reset and this place. Pool.reset marks the outputs as ignored, whereas this makes it possible to reschedule the batches.

Comment (JanuszL) [reply dali/python/nvidia/dali/external_source.py:128]: What if someone calls `cancel_prefetch` while `schedule_and_feed` is running? You can get `self.scheduled_ahead==-1`

**Comment by mzient** [comment]
!build

**Pull request merged**